### DISCUSSION TITLE
How can one apply "vectorized=True" to the Scipy solver?

### DISCUSSION THREAD

**Discussion by HannoSpreeuw** [zwicker-group/py-pde#260] [Q&A]
Take, e.g. the [SIR example](https://py-pde.readthedocs.io/en/latest/examples_gallery/pde_sir.html).
Replace this line
```
sol = eq.solve(state, t_range=50, dt=1e-2, tracker=["progress", tracker])
```
by these
```
solver = ScipySolver(eq)
controller = Controller(solver, t_range=50, tracker=["progress", tracker])
sol = controller.run(state)
```
and it runs just as smooth.
However, when I change that first line to
```
solver = ScipySolver(eq, method = "Radau", vectorized = True)
```
I get
```
ValueError: cannot reshape array of size 9437184 into shape (3,32,32)
```

I am puzzled, I am not sure in what way I should change `evolution_rate` to make this work. Or do I need to change something else?

**Comment by david-zwicker** [comment]
I don't think `vectorize=True` will ever be supported. Have you tried without this option?

**Reply by HannoSpreeuw** [reply]
Yes. That runs without errors, but not very fast. Besides that, the solutions seem numerically unstable, so I think I need to provide a Jacobian.

**Reply by HannoSpreeuw** [reply]
I could speed things up using Numba. If that does not suffice, I will apply `scipy.integrate.solve_ivp` outside `py-pde` and benefit from an additional speedup using `"vectorized=True"`.

**Reply by HannoSpreeuw** [reply]
> Yes. That runs without errors, but not very fast. 

Sorry, that referred to my own code with use of  `solver = ScipySolver(eq, method = "Radau")`.

If I apply "Radau" for the SIR example, it takes "forever".

I thought I could speed it up somewhat using a Jacobian sparsity matrix.
The [solve_ivp documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html#scipy.integrate.solve_ivp) says:
```
jac_sparsityarray_like, sparse matrix or None, optional

    Defines a sparsity structure of the Jacobian matrix for a finite- difference approximation. 
    Its shape must be (n, n).
``` 
So I computed the Jacobian for the SIR example. And found the non-zero elements.
This is the way I ran:
```
jacob_nonzero = np.array([[1,1,0],[1,1,0],[0,1,0]])
solver = ScipySolver(eq, method = "Radau", jac_sparsity = jacob_nonzero)
```
But that gave:
```
operands could not be broadcast together with shapes (3072,) (3,) 
``` 
How should I fix this?
I tried this
```
jac_sparsity = numpy.tile(jacobian_nonzero, 1024)
```
which apparently has compatible dimensions and does not crash, but is still running.
I  am not sure the calculations will be correct.

**Reply by david-zwicker** [reply]
These features of the scipy solver are not supported by `py-pde`. Generally, I only advise to use the scipy solver as a last resort, but I'm actually not aware of any case where it outperforms the explicit solver provided by `py-pde`. Instead, I got best results with the explicit solver with adaptive time stepping. To avoid instabilities you might need to adjust the `tolerance` argument.

**Reply by HannoSpreeuw** [reply]
That is a bummer. I think I need more advanced schemes like "Radau" or "BDF" and provide for an actual Jacobian, because my integrations have shown to be unstable.

**Reply by HannoSpreeuw** [reply]
Perhaps I misinterpreted this comment
```
This class is a thin wrapper around :func:`scipy.integrate.solve_ivp`. In
    particular, it supports all the methods implemented by this function.
```
from the [pde.solvers.scipy source code](https://py-pde.readthedocs.io/en/latest/_modules/pde/solvers/scipy.html#ScipySolver).

On the other hand, it should not be too hard to make Jacobians work within py-pde, right?
Just requires an extra helper for the Jacobian, similar to `rhs_helper` for `fun`, the right-hand side of the system.

**Reply by HannoSpreeuw** [reply]
But thanks for the tip about tweaking the tolerances for the explicit solver. 

For my set of pdes I used
```
solver = ExplicitSolver(eq, scheme="rk", adaptive=True, backend="numba", tolerance=1e-3) 
```
with tolerances between 1e-2 and 1e-6 and this always resulted in
```
RuntimeError: Time step below 1e-10
```
So these equations must be stiff.

**Reply by gnorman7** [reply]
I'm confused by this thread. The source code for the `ScipySolver` has the following, within the `stepper` method:
```
sol = integrate.solve_ivp(
    rhs_helper,
    t_span=(t_start, t_end),
    y0=np.ravel(state.data),
    t_eval=[t_end],  # only store necessary data of the final time point
    **self.solver_params,
)
```

And further, calling solver.solver_params does show that the `kwargs` are being passed properly. Further, trying the input `method='foo'` gives a _SciPy_ error that `'foo'` is not recognized as a solver name. So it seems like it is set up to integrate with SciPy. However, I'm not able to get different behavior by switching the `method` or the tolerances.

**Reply by david-zwicker** [reply]
What do you mean by "I'm not able to get different behavior by switching the method or the tolerances.". What have you tried and what did you expect?

**Reply by gnorman7** [reply]
Nevermind that part, I realize now that I had an error in my code.
